README.md

  [![NPM Version](https://img.shields.io/npm/v/futoin-hkdf.svg?style=flat)](https://www.npmjs.com/package/futoin-hkdf)
  [![NPM Downloads](https://img.shields.io/npm/dm/futoin-hkdf.svg?style=flat)](https://www.npmjs.com/package/futoin-hkdf)
  [![stable](https://img.shields.io/badge/stability-stable-green.svg?style=flat)](https://www.npmjs.com/package/futoin-hkdf)

  [![NPM](https://nodei.co/npm/futoin-hkdf.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/futoin-hkdf/)


# About

Node.js implementation of [RFC5869: HMAC-based Extract-and-Expand Key Derivation Function (HKDF)](https://tools.ietf.org/html/rfc5869).

Additionally, it supports a `HKDF-Expand-Label` variation based on [RFC8446: The Transport Layer Security (TLS) Protocol Version 1.3, section 7.1. Key Schedule](https://datatracker.ietf.org/doc/html/rfc8446#section-7.1).

The implementation is fully compliant with test vectors provided in the RFC.

There are alternative modules, but they are:
* much less performing and/or
* have quite poor code quality at the moment and/or
* are not compliant with RFC (e.g. work only with string parameters) and/or
* not working with current Node.js versions and/or
* do not support arbitrary hash functions and/or
* not reliable dependency for FutoIn Security concept in general.

Standalone HKDF `extract()` and `expand()` actions are also available for advanced usage.

**Documentation** --> [FutoIn Guide](https://futoin.org/docs/miscjs/hkdf/)

Author: [Andrey Galkin](mailto:andrey@futoin.org)

# Performance comparison

The figures in "derived keys per second".

* **futoin-hkdf** - **74 642**
    - fully compliant
* `node-hdkf`/`hdkf` modules - *57 707* (~22% slower)
    - seems to be broken by design
    - **produces wrong results with RFC test vectors**
* `ctrlpanel-hdkf` - *52 181* (~30% slower)
    - seems to be compliant
* `@stablelib/hkdf` - *39 808* (~46% slower)
    - seems to be compliant

# Installation for Node.js

Command line:
```sh
$ npm install futoin-hkdf --save
```
or:

```sh
$ yarn add futoin-hkdf --save
```

# Examples

```javascript
const hkdf = require('futoin-hkdf');

// Parameter overview
//-------------------
// initial keying material
const ikm = 'string-or-buffer';
// required output length in bytes
const length = 16;
// can be empty string or false equivalent
const salt = 'strongly-encouraged';
// optional parameter
const info = 'optional-context';
// HMAC hashing algorithm to use
const hash = 'SHA-256';

// Generic derivation
//-------------------
hkdf(ikm, length, {salt, info, hash}); // Buffer(length) - derived key
hkdf(ikm, length, {salt, info, hash}).toString('hex'); // String(2*length)

// NOTE: all optional paramaters are passed in object

// With some parameters omitted
//-------------------
hkdf(ikm, length, {salt});
hkdf(ikm, length, {info});
hkdf(ikm, length, {hash});
hkdf(ikm, length);

// Advanced usage (only if you know what you are doing)
//-------------------

// As in underlying Node.js crypto library
const lhash = hash.toLowerCase().replace( '-', '' ); // 'sha256'

hkdf.hash_length(lhash); // get hash_len
hkdf.extract(lhash, hash_len, ikm, salt); // run only step #1
hkdf.expand(lhash, hash_len. prk, length, info); // run only step #2

// TLS v1.3+
//-------------------
const hkdf_tls = require('futoin-hkdf/tls');

const label = 'tls13 ...';
const context = Buffer.from( /* E.g some binary hash generation */ '' );

hkdf_tls(ikm, length, {salt, label, context, hash}); // Buffer(length) - derived key

// Advanced usage
hkdf_tls.expand_label(lhash, hash_len. prk, length, labe, context);

// Same as:
hkdf.expand(lhash, hash_len, prk, length, hkdf_tls.info(length, labe, context));

```

# API documentation

## Functions

<dl>
<dt><a href="#hkdf">hkdf(ikm, length, salt, info, hash)</a> ⇒ <code>Buffer</code></dt>
<dd><p>HMAC-based Extract-and-Expand Key Derivation Function (HKDF)</p>
</dd>
<dt><a href="#tls">tls(ikm, length, salt, label, info, hash)</a> ⇒ <code>Buffer</code></dt>
<dd><p>TLS v1.3 HKDF-extract + HKFD-Expand-Label action</p>
</dd>
</dl>

<a name="hkdf"></a>

## hkdf(ikm, length, salt, info, hash) ⇒ <code>Buffer</code>
HMAC-based Extract-and-Expand Key Derivation Function (HKDF)

**Kind**: global function  
**Returns**: <code>Buffer</code> - Raw buffer with derived key of @p length bytes  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| ikm | <code>Buffer</code> \| <code>string</code> |  | Initial Keying Material |
| length | <code>integer</code> |  | Required byte length of output |
| salt | <code>Buffer</code> \| <code>string</code> | <code>&#x27;&#x27;</code> | Optional salt (recommended) |
| info | <code>Buffer</code> \| <code>string</code> | <code>&#x27;&#x27;</code> | Optional context (safe to skip) |
| hash | <code>string</code> | <code>&quot;&#x27;SHA-256&#x27;&quot;</code> | HMAC hash function to use |


* [hkdf(ikm, length, salt, info, hash)](#hkdf) ⇒ <code>Buffer</code>
    * [.hash_length(hash)](#hkdf.hash_length) ⇒ <code>integer</code>
    * [.extract(hash, hash_len, ikm, salt)](#hkdf.extract) ⇒ <code>Buffer</code>
    * [.expand(hash, hash_len, prk, length, info)](#hkdf.expand) ⇒ <code>Buffer</code>

<a name="hkdf.hash_length"></a>

### hkdf.hash\_length(hash) ⇒ <code>integer</code>
Get expected hash length.

**Kind**: static method of [<code>hkdf</code>](#hkdf)  
**Returns**: <code>integer</code> - hash digest byte length  
**Note**: Values are hardcoded with fallback for unknown algorithms.  

| Param | Type | Description |
| --- | --- | --- |
| hash | <code>string</code> | Hash algorithm (as in underlying Node.js crypto library) |

<a name="hkdf.extract"></a>

### hkdf.extract(hash, hash_len, ikm, salt) ⇒ <code>Buffer</code>
HKDF extract action.

**Kind**: static method of [<code>hkdf</code>](#hkdf)  
**Returns**: <code>Buffer</code> - A buffer with pseudorandom key  
**Note**: Values are hardcoded with fallback for unknown algorithms.  

| Param | Type | Description |
| --- | --- | --- |
| hash | <code>string</code> | Hash algorithm (as in underlying Node.js crypto library) |
| hash_len | <code>integer</code> | Hash digest length |
| ikm | <code>Buffer</code> \| <code>string</code> | Initial Keying Material |
| salt | <code>Buffer</code> \| <code>string</code> | Optional salt (recommended) |

<a name="hkdf.expand"></a>

### hkdf.expand(hash, hash_len, prk, length, info) ⇒ <code>Buffer</code>
HKDF expand action.

**Kind**: static method of [<code>hkdf</code>](#hkdf)  
**Returns**: <code>Buffer</code> - A buffer with output keying material  
**Note**: Values are hardcoded with fallback for unknown algorithms.  

| Param | Type | Description |
| --- | --- | --- |
| hash | <code>string</code> | Hash algorithm (as in underlying Node.js crypto library) |
| hash_len | <code>integer</code> | Hash digest length |
| prk | <code>Buffer</code> \| <code>string</code> | A buffer with pseudorandom key |
| length | <code>integer</code> | length of output keying material in octets |
| info | <code>Buffer</code> \| <code>string</code> | Optional context (safe to skip) |

<a name="tls"></a>

## tls(ikm, length, salt, label, info, hash) ⇒ <code>Buffer</code>
TLS v1.3 HKDF-extract + HKFD-Expand-Label action

**Kind**: global function  
**Returns**: <code>Buffer</code> - Raw buffer with derived key of @p length bytes  
**Note**: label and context are limited to 255 bytes!  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| ikm | <code>Buffer</code> \| <code>string</code> |  | Initial Keying Material |
| length | <code>integer</code> |  | Required byte length of output |
| salt | <code>Buffer</code> \| <code>string</code> | <code>&#x27;&#x27;</code> | Optional salt (required by fact) |
| label | <code>Buffer</code> \| <code>string</code> | <code>&#x27;&#x27;</code> | Optional label (required by fact) |
| info | <code>Buffer</code> \| <code>string</code> | <code>&#x27;&#x27;</code> | Optional context (safe to skip) |
| hash | <code>string</code> | <code>&quot;&#x27;SHA-256&#x27;&quot;</code> | HMAC hash function to use |


* [tls(ikm, length, salt, label, info, hash)](#tls) ⇒ <code>Buffer</code>
    * [.info(length, label, context)](#tls.info) ⇒ <code>Buffer</code>
    * [.expand_label(hash, hash_len, prk, length, label, context)](#tls.expand_label) ⇒ <code>Buffer</code>

<a name="tls.info"></a>

### tls.info(length, label, context) ⇒ <code>Buffer</code>
Encode HKDF context parameter in TLS v1.3 style based on RFC8446 TLS v1.3.

**Kind**: static method of [<code>tls</code>](#tls)  
**Returns**: <code>Buffer</code> - A buffer with encoded HKDF context  
**Note**: label and context are limited to 255 bytes!  

| Param | Type | Description |
| --- | --- | --- |
| length | <code>integer</code> | length of output keying material in octets |
| label | <code>string</code> | ASCII label |
| context | <code>Buffer</code> \| <code>string</code> | Binary context or empty string |

<a name="tls.expand_label"></a>

### tls.expand\_label(hash, hash_len, prk, length, label, context) ⇒ <code>Buffer</code>
TLS-HKDF expand label action - a HKDF-Expand-Label variation based on RFC8446 TLS v1.3.

**Kind**: static method of [<code>tls</code>](#tls)  
**Returns**: <code>Buffer</code> - A buffer with output keying material  
**Note**: label and context are limited to 255 bytes!  

| Param | Type | Description |
| --- | --- | --- |
| hash | <code>string</code> | Hash algorithm (as in underlying Node.js crypto library) |
| hash_len | <code>integer</code> | Hash digest length |
| prk | <code>Buffer</code> \| <code>string</code> | A buffer with pseudorandom key |
| length | <code>integer</code> | length of output keying material in octets |
| label | <code>string</code> | ASCII label |
| context | <code>Buffer</code> \| <code>string</code> | Binary context or empty string |




CHANGELOG.txt
=== 1.5.3 (2023-07-30) ===
CHANGED: dependency maintenance

=== 1.5.2 (2023-03-27) ===
CHANGED: dependency maintenance

=== 1.5.1 (2022-05-27) ===
CHANGED: dependency maintenance
FIXED: minor fixes to README

=== 1.5.0 (2022-01-24) ===
CHANGED: use Buffer via module to aid usage in browser environment (thanks to johnwchadwick)
CHANGED: dependency maintenance

=== 1.4.3 (2021-12-19) ===
CHANGED: dependency maintenance

=== 1.4.2 (2021-07-05) ===
FIXED: to include tls.* files in the package

=== 1.4.1 (2021-07-02) ===
FIXED: hkdf.tls() direct invocation

=== 1.4.0 (2021-07-02) ===
CHANGED: hdkf.expand() info parameter to use binary buffer as-is without copying
CHANGED: dependency maintenance
NEW: tls, tls.info() and tls.expand_label() feature inspired by TLS v1.3 RFC8446 section 7.1

=== 1.3.3 (2021-01-31) ===
CHANGED: dependency maintenance

=== 1.3.2 (2020-03-15) ===
CHANGED: dependency maintenance

=== 1.3.1 (2020-01-10) ===
FIXED: clarified hash parameter for advanced usage

=== 1.3.0 (2019-12-24) ===
CHANGED: fixed TypeScript definitions
CHANGED: updated deps

=== 1.2.1 (2019-09-27) ===
CHANGED: deps maintenance

=== 1.2.0 (2019-08-01) ===
NEW: TypeScript bindings

=== 1.1.0 (2019-07-14) ===
CHANGED: dependency maintenance
CHANGED: dropped Node v6 support
NEW: hardcoded SHA3 lengths and length caching for unknown algos

=== 1.0.3 (2019-01-11) ===
CHANGED: dependency maintenance

=== 1.0.2 (2018-08-11) ===
CHANGED: dependency maintenance

=== 1.0.1 (2018-04-23) ===
FIXED: to allow maximum length of 255 steps for ORM
    - Thanks to https://github.com/davidjludwig
      for reporting https://github.com/futoin/util-js-hkdf/issues/1
      and providing a fix.

=== 1.0.0 (2018-01-12) ===
No changes. Going final.

=== 0.9.0 (2018-01-12) ===
Initial version


test/unittest.js
'use strict';

const { Buffer } = require( 'buffer' );
const crypto = require( 'crypto' );
const hkdf = require( '../hkdf' );
const tls = require( '../tls' );
const expect = require( 'chai' ).expect;

describe( 'HKDF', function() {
    it ( 'use proper hash lengths', function() {
        const algos = [
            'sha256',
            'sha512',
            'sha224',
            'sha384',
            // 'blake2s256',
            // 'blake2b512',
            'sha1',
            'md5',
            // 'gost',
        ];

        for ( let a of algos ) {
            expect( hkdf.hash_length( a ) ).to.equal( crypto.createHash( a ).digest().length );
        }

        const future_algos = {
            'sha3-256' : 32,
            'sha3-512' : 64,
            'sha3-224' : 28,
            'sha3-384' : 48,
            blake2s256 : 32,
            blake2b512 : 64,
        };

        for ( let a in future_algos ) {
            expect( hkdf.hash_length( a ) ).to.equal( future_algos[a] );
        }
    } );

    it ( 'use hash length caching', function() {
        const algos = [
            'SHA256',
            'RIPEMD160',
            'SHA256',
            'RIPEMD160',
        ];

        for ( let a of algos ) {
            expect( hkdf.hash_length( a ) ).to.equal( crypto.createHash( a ).digest().length );
        }
    } );

    it ( 'should correctly handle optional parameters', function() {
        // note, values are derived from this functionality - can be invalid in theory
        expect( hkdf( 'abc', 12 ).toString( 'hex' ) ).to.equal( '4af82925ee74ef036c1ff38e' );
        expect( hkdf( 'abc', 12, { salt:'' } ).toString( 'hex' ) ).to.equal( '4af82925ee74ef036c1ff38e' );
        expect( hkdf( 'abc', 12, { salt:'123' } ).toString( 'hex' ) ).to.equal( '2385d8d57460fb520e05cb53' );
        expect( hkdf( 'abc', 12, { info:'' } ).toString( 'hex' ) ).to.equal( '4af82925ee74ef036c1ff38e' );
        expect( hkdf( 'abc', 12, { info:'123' } ).toString( 'hex' ) ).to.equal( '9fd198216181de061d18c424' );
        expect( hkdf( 'abc', 12, { hash:'MD5' } ).toString( 'hex' ) ).to.equal( '108ab4ed0e6336264c1f7182' );
    } );

    // as per https://tools.ietf.org/html/rfc6234
    //    *      okm_len: [in]
    //    *          The length of the buffer to hold okm.
    //    *          okm_len must be <= 255 * USHABlockSize(whichSha)
    // --------------------
    // 255 * 32 = 8160    so 8160 is a valid case for sha256 (and 8161 should be rejected)

    it ( 'should detect too long OKM issue', function() {
        expect( () => {
            hkdf( 'abc', 8161 );
        } ).to.throw( 'OKM length 8161 is too long for sha256 hash' );
    } );

    it ( 'should fallback hash length for unknown algorithms', function() {
        expect( () => {
            hkdf.hash_length( 'invalid' );
        } ).to.throw( 'Digest method not supported' );
    } );

    // NOTE: source for test vectors - https://tools.ietf.org/html/rfc5869
    const test_vectors = {
        1: {
            Hash: 'SHA-256',
            IKM: '0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b',
            salt: '000102030405060708090a0b0c',
            info: 'f0f1f2f3f4f5f6f7f8f9',
            L: 42,
            PRK: [
                '077709362c2e32df0ddc3f0dc47bba63',
                '90b6c73bb50f9c3122ec844ad7c2b3e5',
            ].join( '' ),
            OKM: [
                '3cb25f25faacd57a90434f64d0362f2a',
                '2d2d0a90cf1a5a4c5db02d56ecc4c5bf',
                '34007208d5b887185865',
            ].join( '' ),
        },

        2: {
            Hash: 'SHA-256',
            IKM: [
                '000102030405060708090a0b0c0d0e0f',
                '101112131415161718191a1b1c1d1e1f',
                '202122232425262728292a2b2c2d2e2f',
                '303132333435363738393a3b3c3d3e3f',
                '404142434445464748494a4b4c4d4e4f',
            ].join( '' ),
            salt: [
                '606162636465666768696a6b6c6d6e6f',
                '707172737475767778797a7b7c7d7e7f',
                '808182838485868788898a8b8c8d8e8f',
                '909192939495969798999a9b9c9d9e9f',
                'a0a1a2a3a4a5a6a7a8a9aaabacadaeaf',
            ].join( '' ),
            info: [
                'b0b1b2b3b4b5b6b7b8b9babbbcbdbebf',
                'c0c1c2c3c4c5c6c7c8c9cacbcccdcecf',
                'd0d1d2d3d4d5d6d7d8d9dadbdcdddedf',
                'e0e1e2e3e4e5e6e7e8e9eaebecedeeef',
                'f0f1f2f3f4f5f6f7f8f9fafbfcfdfeff',
            ].join( '' ),
            L: 82,
            PRK: [
                '06a6b88c5853361a06104c9ceb35b45c',
                'ef760014904671014a193f40c15fc244',
            ].join( '' ),
            OKM: [
                'b11e398dc80327a1c8e7f78c596a4934',
                '4f012eda2d4efad8a050cc4c19afa97c',
                '59045a99cac7827271cb41c65e590e09',
                'da3275600c2f09b8367793a9aca3db71',
                'cc30c58179ec3e87c14c01d5c1f3434f',
                '1d87',
            ].join( '' ),
        },

        3: {
            Hash: 'SHA-256',
            IKM: '0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b',
            salt: '',
            info: '',
            L: 42,
            PRK: [
                '19ef24a32c717b167f33a91d6f648bdf',
                '96596776afdb6377ac434c1c293ccb04',
            ].join( '' ),
            OKM: [
                '8da4e775a563c18f715f802a063c5a31',
                'b8a11f5c5ee1879ec3454e5f3c738d2d',
                '9d201395faa4b61a96c8',
            ].join( '' ),
        },

        4: {
            Hash: 'SHA-1',
            IKM: '0b0b0b0b0b0b0b0b0b0b0b',
            salt: '000102030405060708090a0b0c',
            info: 'f0f1f2f3f4f5f6f7f8f9',
            L: 42,
            PRK: '9b6c18c432a7bf8f0e71c8eb88f4b30baa2ba243',
            OKM: [
                '085a01ea1b10f36933068b56efa5ad81',
                'a4f14b822f5b091568a9cdd4f155fda2',
                'c22e422478d305f3f896',
            ].join( '' ),
        },

        5: {
            Hash: 'SHA-1',
            IKM: [
                '000102030405060708090a0b0c0d0e0f',
                '101112131415161718191a1b1c1d1e1f',
                '202122232425262728292a2b2c2d2e2f',
                '303132333435363738393a3b3c3d3e3f',
                '404142434445464748494a4b4c4d4e4f',
            ].join( '' ),
            salt: [
                '606162636465666768696a6b6c6d6e6f',
                '707172737475767778797a7b7c7d7e7f',
                '808182838485868788898a8b8c8d8e8f',
                '909192939495969798999a9b9c9d9e9f',
                'a0a1a2a3a4a5a6a7a8a9aaabacadaeaf',
            ].join( '' ),
            info: [
                'b0b1b2b3b4b5b6b7b8b9babbbcbdbebf',
                'c0c1c2c3c4c5c6c7c8c9cacbcccdcecf',
                'd0d1d2d3d4d5d6d7d8d9dadbdcdddedf',
                'e0e1e2e3e4e5e6e7e8e9eaebecedeeef',
                'f0f1f2f3f4f5f6f7f8f9fafbfcfdfeff',
            ].join( '' ),
            L: 82,
            PRK: '8adae09a2a307059478d309b26c4115a224cfaf6',
            OKM: [
                '0bd770a74d1160f7c9f12cd5912a06eb',
                'ff6adcae899d92191fe4305673ba2ffe',
                '8fa3f1a4e5ad79f3f334b3b202b2173c',
                '486ea37ce3d397ed034c7f9dfeb15c5e',
                '927336d0441f4c4300e2cff0d0900b52',
                'd3b4',
            ].join( '' ),
        },

        6: {
            Hash: 'SHA-1',
            IKM: '0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b',
            salt: '',
            info: '',
            L: 42,
            PRK: 'da8c8a73c7fa77288ec6f5e7c297786aa0d32d01',
            OKM: [
                '0ac1af7002b3d761d1e55298da9d0506',
                'b9ae52057220a306e07b6b87e8df21d0',
                'ea00033de03984d34918',
            ].join( '' ),
        },

        7: {
            Hash: 'SHA-1',
            IKM: '0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c',
            salt: null,
            info: '',
            L: 42,
            PRK: '2adccada18779e7c2077ad2eb19d3f3e731385dd',
            OKM: [
                '2c91117204d745f3500d636a62f64f0a',
                'b3bae548aa53d423b0d1f27ebba6f5e5',
                '673a081d70cce7acfc48',
            ].join( '' ),
        },
    };

    for ( let k in test_vectors ) {
        const v = test_vectors[k];

        it ( `should pass RFC Test Vector #${k}`, function() {
            const ihash = v.Hash.toLowerCase().replace( '-', '' );

            expect( hkdf.extract(
                ihash,
                hkdf.hash_length( ihash ),
                Buffer.from( v.IKM, 'hex' ),
                v.salt && Buffer.from( v.salt, 'hex' )
            ).toString( 'hex' ) ).to.equal( v.PRK );

            expect( hkdf(
                Buffer.from( v.IKM, 'hex' ),
                v.L,
                {
                    salt: v.salt && Buffer.from( v.salt, 'hex' ),
                    info: Buffer.from( v.info, 'hex' ),
                    hash: v.Hash,
                }
            ).toString( 'hex' ) ).to.equal( v.OKM );
        } );
    }
} );

describe( 'QUIC-TLS', function() {
    // Based on test vectors from https://datatracker.ietf.org/doc/html/rfc9001

    // 5.2.  Initial Secrets
    const initial_salt = Buffer.from( '38762cf7f55934b34d179ae6a4c80cadccbb7f0a', 'hex' );
    const algo = 'sha256';
    const hash_len = hkdf.hash_length( algo );

    // Appendix A.  Sample Packet Protection
    const dcid = Buffer.from( '8394c8f03e515708', 'hex' );

    const initial_secret = hkdf.extract( algo, hash_len, dcid, initial_salt );

    it ( 'should generate correct initial_secret', () => {
        expect( initial_secret.toString( 'hex' ) ).to.equal( '7db5df06e7a69e432496adedb00851923595221596ae2ae9fb8115c1e9ed0a44' );
    } );

    it ( 'should generate correct client secrets', () => {
        const client_initial_secret = tls.expand_label( algo, hash_len, initial_secret, 32, 'tls13 client in', '' );
        expect( client_initial_secret.toString( 'hex' ) )
            .to.equal( 'c00cf151ca5be075ed0ebfb5c80323c42d6b7db67881289af4008f1f6c357aea' );

        // key
        expect( tls.expand_label( algo, hash_len, client_initial_secret, 16, 'tls13 quic key', '' ).toString( 'hex' ) )
            .to.equal( '1f369613dd76d5467730efcbe3b1a22d' );

        // iv
        expect( tls.expand_label( algo, hash_len, client_initial_secret, 12, 'tls13 quic iv', '' ).toString( 'hex' ) )
            .to.equal( 'fa044b2f42a3fd3b46fb255c' );

        // hp
        expect( tls.expand_label( algo, hash_len, client_initial_secret, 16, 'tls13 quic hp', '' ).toString( 'hex' ) )
            .to.equal( '9f50449e04a0e810283a1e9933adedd2' );
    } );

    it ( 'should generate correct server secrets', () => {
        const server_initial_secret = tls.expand_label( algo, hash_len, initial_secret, 32, 'tls13 server in', '' );
        expect( server_initial_secret.toString( 'hex' ) )
            .to.equal( '3c199828fd139efd216c155ad844cc81fb82fa8d7446fa7d78be803acdda951b' );

        // key
        expect( tls.expand_label( algo, hash_len, server_initial_secret, 16, 'tls13 quic key', '' ).toString( 'hex' ) )
            .to.equal( 'cf3a5331653c364c88f0f379b6067e37' );

        // iv
        expect( tls.expand_label( algo, hash_len, server_initial_secret, 12, 'tls13 quic iv', '' ).toString( 'hex' ) )
            .to.equal( '0ac1493ca1905853b0bba03e' );

        // hp
        expect( tls.expand_label( algo, hash_len, server_initial_secret, 16, 'tls13 quic hp', '' ).toString( 'hex' ) )
            .to.equal( 'c206b8d9b9f0f37644430b490eeaa314' );
    } );

    it ( 'should work via all-in-one call', () => {
        expect( tls( dcid, 32, { salt: initial_salt, label: 'tls13 client in' } ).toString( 'hex' ) )
            .to.equal( 'c00cf151ca5be075ed0ebfb5c80323c42d6b7db67881289af4008f1f6c357aea' );
    } );
} );


test/benchmark.js
'use strict';

const assert = require( 'assert' );
const { Suite } = require( 'benchmark' );
const { Buffer } = require( 'buffer' );

const futoin_hkdf = require( '../hkdf' );

// NOTE: requires patching to work with latest Node.js
//const OtherHKDF = require( './foreign/hkdf' );

//const ctrlpanel = require( './foreign/ctrlpanel' );
//const stablelib = require( '@stablelib/hkdf' ).HKDF;
//const stablelib_sha256 = require( '@stablelib/sha256' ).SHA256;

{
    const hash = 'SHA-256';
    const l_hash = 'sha256';
    const ikm = Buffer.from( '0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b', 'hex' );
    const s_ikm = ikm.toString( 'ascii' );
    const salt = Buffer.from( '000102030405060708090a0b0c', 'hex' );
    const s_salt = ikm.toString( 'ascii' );
    const info = Buffer.from( 'f0f1f2f3f4f5f6f7f8f9', 'hex' );
    const s_info = ikm.toString( 'ascii' );
    const L = 42;
    const okm = Buffer.from( [
        '3cb25f25faacd57a90434f64d0362f2a',
        '2d2d0a90cf1a5a4c5db02d56ecc4c5bf',
        '34007208d5b887185865',
    ].join( '' ), 'hex' );
    const s_okm = okm.toString( 'ascii' );

    const bsha256 = new Suite( 'HKDF SHA-256' );
    bsha256
        .add( 'FutoIn', function() {
            const res = futoin_hkdf( ikm, L, { salt, info, hash } );
            //assert( res.equals( okm ) );
        } )
        /*.add('Other HKDF', {
            defer: true,
            fn: function(deferred){
                const hkdf = new OtherHKDF(l_hash, salt, ikm);
                hkdf.derive( info, L, (res) => {
                    // FAILS
                    //assert( res.equals( okm ) );
                    deferred.resolve();
                } );
            }
        })
        .add('CtrlPanel', {
            defer: true,
            fn: function(deferred){
                ctrlpanel( salt, ikm, info, L, hash )
                .then( (res) => {
                    //assert( Buffer.from(res).equals( okm ) );
                    deferred.resolve();
                } );
            }
        })
        .add('StableLib', function(){
            const hkdf = new stablelib(
                stablelib_sha256,
                ikm,
                salt,
                info,
            );
            const res = hkdf.expand( L );
            //assert( Buffer.from(res).equals( okm ) );
        })
        .on('complete', function() {
            console.log(this);
        })*/
        .run();
}


hkdf.js
'use strict';

/**
 * @file
 *
 * Copyright 2018 FutoIn Project (https://futoin.org)
 * Copyright 2018 Andrey Galkin <andrey@futoin.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const { createHash, createHmac } = require( 'crypto' );

const g_digestLenCache = {};

/**
 * Get expected hash length.
 *
 * @func
 * @alias hkdf.hash_length
 * @param {string} hash - Hash algorithm
 * @returns {integer} hash digest byte length
 *
 * @note Values are hardcoded with fallback for unknown algorithms.
 */
const hash_length = ( hash ) => {
    switch ( hash ) {
    case 'sha256': return 32;
    case 'sha512': return 64;
    case 'sha224': return 28;
    case 'sha384': return 48;
    case 'sha3-256': return 32;
    case 'sha3-512': return 64;
    case 'sha3-224': return 28;
    case 'sha3-384': return 48;
    case 'blake2s256': return 32;
    case 'blake2b512': return 64;
    case 'sha1': return 20;
    case 'md5': return 16;
    default: {
        let len = g_digestLenCache[hash];

        if ( len === undefined ) {
            len = createHash( hash ).digest().length;
            g_digestLenCache[hash] = len;
        }

        return len;
    }
    }
};

/**
 * HKDF extract action.
 *
 * @func
 * @alias hkdf.extract
 * @param {string} hash - Hash algorithm
 * @param {integer} hash_len - Hash digest length
 * @param {Buffer|string} ikm - Initial Keying Material
 * @param {Buffer|string} salt - Optional salt (recommended)
 * @returns {Buffer} A buffer with pseudorandom key
 *
 * @note Values are hardcoded with fallback for unknown algorithms.
 */
const hkdf_extract = ( hash, hash_len, ikm, salt ) => {
    const b_ikm = Buffer.isBuffer( ikm ) ? ikm : Buffer.from( ikm );
    const b_salt = ( salt && salt.length ) ? Buffer.from( salt ) : Buffer.alloc( hash_len, 0 );

    return createHmac( hash, b_salt ).update( b_ikm ).digest();
};

/**
 * HKDF expand action.
 *
 * @func
 * @alias hkdf.expand
 * @param {string} hash - Hash algorithm
 * @param {integer} hash_len - Hash digest length
 * @param {Buffer|string} prk - A buffer with pseudorandom key
 * @param {integer} length - length of output keying material in octets
 * @param {Buffer|string} info - Optional context (safe to skip)
 * @returns {Buffer} A buffer with output keying material
 *
 * @note Values are hardcoded with fallback for unknown algorithms.
 */
const hkdf_expand = ( hash, hash_len, prk, length, info ) => {
    const b_info = Buffer.from( info || '' );
    const info_len = b_info.length;

    const steps = Math.ceil( length / hash_len );

    if ( steps > 0xFF ) {
        throw new Error( `OKM length ${length} is too long for ${hash} hash` );
    }

    // use single buffer with unnecessary create/copy/move operations
    const t = Buffer.alloc( hash_len * steps + info_len + 1 );

    for ( let c = 1, start = 0, end = 0; c <= steps; ++c ) {
        // add info
        b_info.copy( t, end );
        // add counter
        t[ end + info_len ] = c;

        createHmac( hash, prk )
            // use view: T(C) = T(C-1) | info | C
            .update( t.slice( start, end + info_len + 1 ) )
            .digest()
            // put back to the same buffer
            .copy( t, end );

        start = end; // used for T(C-1) start
        end += hash_len; // used for T(C-1) end & overall end
    }

    return t.slice( 0, length );
};

/**
 * HMAC-based Extract-and-Expand Key Derivation Function (HKDF)
 *
 * @param {Buffer|string} ikm - Initial Keying Material
 * @param {integer} length - Required byte length of output
 * @param {Buffer|string} salt='' - Optional salt (recommended)
 * @param {Buffer|string} info='' - Optional context (safe to skip)
 * @param {string} hash='SHA-256' - HMAC hash function to use
 * @returns {Buffer} Raw buffer with derived key of @p length bytes
 */
function hkdf( ikm, length, { salt='', info='', hash='SHA-256' } = {} ) {
    hash = hash.toLowerCase().replace( '-', '' );

    // 0. Hash length
    const hash_len = hash_length( hash );

    // 1. extract
    const prk = hkdf_extract( hash, hash_len, ikm, salt );

    // 2. expand
    return hkdf_expand( hash, hash_len, prk, length, info );
}

Object.defineProperties( hkdf, {
    hash_length : {
        configurable: false,
        enumerable: false,
        writable: false,
        value: hash_length,
    },
    extract : {
        configurable: false,
        enumerable: false,
        writable: false,
        value: hkdf_extract,
    },
    expand : {
        configurable: false,
        enumerable: false,
        writable: false,
        value: hkdf_expand,
    },
} );

module.exports = hkdf;
